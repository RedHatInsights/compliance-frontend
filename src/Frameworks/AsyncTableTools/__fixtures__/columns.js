import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
    key: 'id',
    manageable: false,
  },
  {
    title: 'Name',
    key: 'id',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    key: 'another_name_column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    key: 'desc',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Download button',
    key: 'download',
    manageable: false,
    export: false,
    renderFunc: () => 'download link',
  },
];
