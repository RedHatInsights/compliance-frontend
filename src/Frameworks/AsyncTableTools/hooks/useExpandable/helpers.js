import React from 'react';

const detailsRowForRule = (item, DetailsComponent, colSpan, runningIndex) => ({
  parent: runningIndex() - 1,
  props: {
    ...item.props,
    'aria-setsize': 0,
  },
  cells: [
    {
      title: <DetailsComponent item={item} key={'item-' + item.rowId} />,
      props: {
        ...(colSpan ? { colSpan } : {}),
        // TODO This removes the checkbox, however this should maybe be fixed differently
        className: 'compliance-rule-details',
      },
    },
  ],
});

export const itemDetailsRow = (item, options, runningIndex) =>
  typeof options?.detailsComponent !== 'undefined' &&
  detailsRowForRule(
    item,
    options.detailsComponent,
    options.columns?.length,
    runningIndex
  );

const expandTreeTableRow = (firstRow, isOpen) => ({
  ...firstRow,
  props: {
    ...(firstRow.props || {}),
    isExpanded: isOpen,
  },
});

const expandTableRow = (firstRow, isOpen) => ({
  ...firstRow,
  isOpen,
});

export const addExpandProp = (firstRow, isTreeTable, isOpen) =>
  isTreeTable
    ? expandTreeTableRow(firstRow, isOpen)
    : expandTableRow(firstRow, isOpen);
