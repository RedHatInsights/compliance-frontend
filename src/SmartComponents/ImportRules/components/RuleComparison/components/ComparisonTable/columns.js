import React from 'react';
import { nowrap } from '@patternfly/react-table';

import RuleCell from './components/RuleCell';
import ProfileVersionCell from './components/ProfileVersionCell';

const rule = {
  title: <strong>Rule</strong>,
  Component: RuleCell,
};

export const profileVersion = (version, targetVersion, additionalProps) => ({
  dataLabel: `RHEL ${version.os_major_version}.${version.os_minor_version} -  SSG version ${version.ssg_version}`,
  title: (
    <>
      <strong>
        RHEL {version.os_major_version}.{version.os_minor_version}
      </strong>
      <br />
      SSG version {version.ssg_version}
    </>
  ),
  Component: (props) => (
    <ProfileVersionCell
      {...{
        version,
        targetVersion,
        ...props,
        ...additionalProps,
      }}
    />
  ),
  transforms: [nowrap],
});

export default [rule];
