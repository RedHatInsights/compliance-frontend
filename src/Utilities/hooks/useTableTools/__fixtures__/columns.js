import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        sortByProperty: 'name',
        renderFunc: (_a, _b, item) => (
            item.name
        )
    }, {
        title: 'Description',
        sortByProperty: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (item.name),
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }
];
