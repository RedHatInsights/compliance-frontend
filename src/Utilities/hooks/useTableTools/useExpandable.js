import React, { useCallback, useMemo, useState } from 'react';
import { findParentsForItemInTree } from './rowBuilderHelpers';

const calculateColSpan = (columns, options) => {
  let colSpan = columns.length + 1;
  if (options.selectable || options.hasRadioSelect) {
    colSpan++;
  }
  return colSpan;
};

const childRowForRule = (item, idx, DetailsComponent, colSpan) => ({
  parent: idx * 2,
  fullWidth: true,
  cells: [
    {
      title: <DetailsComponent item={item} key={'item-' + item.rowId} />,
      props: { colSpan },
    },
  ],
});

const itemDetailsRow = (item, idx, options, columns) =>
  typeof options?.detailsComponent !== 'undefined' &&
  childRowForRule(
    item,
    idx,
    options.detailsComponent,
    calculateColSpan(columns, options)
  );

const useExpandable = (options) => {
  const enableExpanbale = !!options.detailsComponent && !options.showTreeTable;
  const findParentsForItem = useMemo(
    () => findParentsForItemInTree(options.tableTree),
    [options.tableTreee]
  );

  const [openItems, setOpenItems] = useState([]);
  const onCollapse = useCallback(
    (_event, _index, _isOpen, row) => {
      const parentItemIds = findParentsForItem(row.itemId);

      if (openItems.includes(row.itemId)) {
        setOpenItems((currentOpenItems) =>
          currentOpenItems.filter((itemId) => itemId !== row.itemId)
        );
      } else {
        setOpenItems((currentOpenItems) => [
          ...currentOpenItems,
          ...parentItemIds,
          row.itemId,
        ]);
      }
    },
    [openItems, setOpenItems]
  );

  const openItem = (row, item, columns, rowIndex) => {
    const isOpen = openItems.includes(item.itemId);
    const newRow = {
      ...row,
      isOpen,
    };
    const expandableRow = itemDetailsRow(item, rowIndex, options, columns);

    return [newRow, expandableRow];
  };

  return enableExpanbale
    ? {
        transformer: openItem,
        tableProps: {
          onCollapse,
        },
        openItems,
      }
    : {
        tableProps: {
          onCollapse,
        },
        openItems,
      };
};

export default useExpandable;
