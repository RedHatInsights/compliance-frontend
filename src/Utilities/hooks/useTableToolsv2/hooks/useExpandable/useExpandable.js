import useSelectionManager from '../useSelectionManager';

// NOTE: This expanding is, i believe only used in the list view of the RulesTable
/**
 * Provides props for a Patternfly table to manage expandable items/rows.
 *
 */
const useExpandable = () => {
  const { selection: openItems, toggle } = useSelectionManager([]);

  const onCollapse = (_event, _index, _isOpen, { itemId }) => toggle(itemId);

  // TODO This function will be passed as a "rowTransformer" to the `rowsBuilder`, to "open" a row
  // This means it needs to check if the id of the current row is "selected" and also render the "DetailsRow" component as an additional row
  const openItem = (row) => {};

  return {
    openItem,
    openItems,
    tableProps: {
      onCollapse,
    },
  };
};

export default useExpandable;
