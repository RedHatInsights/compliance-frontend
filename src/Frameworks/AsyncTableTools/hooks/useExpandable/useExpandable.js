import { useCallback, useEffect } from 'react';
import isEqual from 'lodash/isEqual';
import useSelectionManager from '../useSelectionManager';
import useTableState from '../useTableState';
import { itemDetailsRow } from './helpers';

/**
 *  @typedef {object | undefined} useExpandableReturn
 *
 *  @property {Function} [openItem]   "Transformer" function to be passed to the {@link rowsBuilder}
 *  @property {object}   [tableProps] Object containing Patternfly (deprecated) Table props
 *
 */

/**
 * Provides props for a Patternfly table to manage expandable items/rows.
 *
 *  @param   {object}              [options]                  AsyncTableTools options
 *  @param   {object}              [options.detailsComponent] A component that should be rendered as a details row
 *
 *  @returns {useExpandableReturn}                            An object of props meant to be used in the {@link AsyncTableToolsTable}
 *
 *  @category AsyncTableTools
 *  @subcategory Hooks
 *
 */
const useExpandable = (options) => {
  const enableExpandingRow = !!options?.detailsComponent || !!options.treeTable;
  const { selection: openItems, toggle } = useSelectionManager([]);
  // TODO If the selection manager is based on `useTableState`, observes can be used to reset open items
  const [openItemsState, setOpenItemsState] = useTableState('open-items');

  const onCollapse = (_event, _index, _isOpen, { itemId }) => {
    toggle(itemId);
  };

  const openItem = useCallback(
    (row, _selectedIds, index) => {
      const isOpen = (openItems || []).includes(row.itemId);

      return isOpen
        ? [
            {
              ...row,
              isOpen,
            },
            itemDetailsRow(row, index, options),
          ]
        : [{ ...row, isOpen }];
    },
    [openItems, options]
  );

  // TODO This is hackish. We should rather have a selection manager based on a table state
  useEffect(() => {
    !isEqual(openItems, openItemsState) && setOpenItemsState(openItems);
  }, [openItems, openItemsState, setOpenItemsState]);

  return enableExpandingRow
    ? {
        openItems,
        openItem,
        tableProps: {
          onCollapse,
        },
        tableView: {
          onCollapse,
          openItems,
        },
      }
    : {};
};

export default useExpandable;
