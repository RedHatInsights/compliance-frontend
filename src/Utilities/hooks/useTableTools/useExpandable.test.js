import { renderHook, act } from '@testing-library/react-hooks';
import useExpandable from './useExpandable';
import items from './__fixtures__/items';
import columns from './__fixtures__/columns';

const ExampleDetailsRow = (item) => {
  return `DETAILS for ${item.name}`;
};

describe('useExpandable', () => {
  const exampleItems = items(30);
  const defaultOptions = {
    detailsComponent: ExampleDetailsRow,
  };

  it('returns an expandable configuration', () => {
    const { result } = renderHook(() => useExpandable(defaultOptions));

    expect(result).toMatchSnapshot();
  });

  it('returns a transformer to open items', () => {
    const { result } = renderHook(() => useExpandable(defaultOptions));
    const openedItem = result.current.transformer(
      exampleItems[0],
      {
        ...exampleItems[0],
        itemId: 1,
      },
      columns,
      1
    );

    expect(openedItem).toMatchSnapshot();
  });

  it('sets an item as open onCollapse', () => {
    const { result } = renderHook(() => useExpandable(defaultOptions));

    act(() => {
      result.current.tableProps.onCollapse(undefined, undefined, undefined, {
        itemId: 1,
      });
    });

    const openedItem = result.current.transformer(
      exampleItems[0],
      {
        ...exampleItems[0],
        itemId: 1,
      },
      columns,
      1
    );

    expect(openedItem).toMatchSnapshot();
  });

  it('should not return an expandable configuration with treetable active', () => {
    const { result } = renderHook(() =>
      useExpandable({
        ...defaultOptions,
        tableTree: ['item'],
      })
    );

    expect(Object.keys(result.current).length).toEqual(0);
  });
});
