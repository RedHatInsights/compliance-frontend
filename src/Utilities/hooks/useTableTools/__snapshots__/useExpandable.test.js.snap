// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useExpandable returns a transformer to open items 1`] = `
Array [
  Object {
    "description": "DESCRIPTION",
    "id": 1,
    "isOpen": false,
    "name": "TEST ITEM #1",
    "severity": "medium",
  },
  Object {
    "cells": Array [
      Object {
        "props": Object {
          "colSpan": 7,
        },
        "title": <ExampleDetailsRow
          item={
            Object {
              "description": "DESCRIPTION",
              "id": 1,
              "itemId": 1,
              "name": "TEST ITEM #1",
              "severity": "medium",
            }
          }
        />,
      },
    ],
    "fullWidth": true,
    "parent": 2,
  },
]
`;

exports[`useExpandable returns an expandable configuration 1`] = `
Object {
  "all": Array [
    Object {
      "openItems": Array [],
      "tableProps": Object {
        "onCollapse": [Function],
      },
      "transformer": [Function],
    },
  ],
  "current": Object {
    "openItems": Array [],
    "tableProps": Object {
      "onCollapse": [Function],
    },
    "transformer": [Function],
  },
  "error": undefined,
}
`;

exports[`useExpandable sets an item as open onCollapse 1`] = `
Array [
  Object {
    "description": "DESCRIPTION",
    "id": 1,
    "isOpen": true,
    "name": "TEST ITEM #1",
    "severity": "medium",
  },
  Object {
    "cells": Array [
      Object {
        "props": Object {
          "colSpan": 7,
        },
        "title": <ExampleDetailsRow
          item={
            Object {
              "description": "DESCRIPTION",
              "id": 1,
              "itemId": 1,
              "name": "TEST ITEM #1",
              "severity": "medium",
            }
          }
        />,
      },
    ],
    "fullWidth": true,
    "parent": 2,
  },
]
`;
